### PR TITLE
Add Router Dispatcher component

### PR DESCRIPTION
**Pull request by matks**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Description?  | Add Router Dispatcher component page
| Fixed ticket? | Fixes https://github.com/PrestaShop/docs/issues/939

### CODE REVIEW

**Comment by matks** [comment]
Could you please review this work @PrestaShop/prestashop-maintainers  😉

File: src/content/1.7/development/components/router-dispatcher.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+menuTitle: Router and Dispatcher
+---
+
+# Router and Dispatcher
+
+The legacy Dispatcher and the Symfony Router are in charge of parsing incoming HTTP requests and finding the right Controller to return HTTP response.
+
+### Front office routing
+
+The Front office controllers lie in directory `controllers/front`.
```
Comment (kpodemski) [inline src/content/1.7/development/components/router-dispatcher.md:12]:
```suggestion
The Front office controllers lie in the `controllers/front` directory.
```

File: src/content/1.7/development/components/router-dispatcher.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+The Dispatcher match a given HTTP request with a Front Office controller.
+
+When you visit `/` on your shop you are being returned the Homepage.
```
Comment (kpodemski) [inline src/content/1.7/development/components/router-dispatcher.md:16]:
```suggestion
When you visit `/` on your shop you are being returned the Homepage, unless you have your shop installed in a subdirectory, then `/your_subdirectory` is your Homepage.
```

File: src/content/1.7/development/components/router-dispatcher.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+What happened behind the scenes is that
```
Comment (kpodemski) [inline src/content/1.7/development/components/router-dispatcher.md:18]:
```suggestion
What happened behind the scenes is that:
```

File: src/content/1.7/development/components/router-dispatcher.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+If you send another HTTP request, the same sequence will happen
+- Dispatcher finds the right Front controller
+- Front controller builds the right response using a Smarty template and relying on PrestaShop business logic
+
+## Back office routing
+
+The Back Office is split between the legacy part and the migrated part.
+
+The legacy backend relies on controllers you can find in directory `controllers/admin`.
```
Comment (kpodemski) [inline src/content/1.7/development/components/router-dispatcher.md:32]:
```suggestion
The legacy backend relies on controllers you can find in the `controllers/admin` directory.
```

File: src/content/1.7/development/components/router-dispatcher.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+The Symfony backend relies on controllers you can find in `src/PrestaShopBundle/Controller/Admin`.
+
+The Back Offie routing match a given HTTP request with a Controller.
+
+### Legacy Back Office routing
+
+When you visit `/admin{x}/index.php?controller=AdminCarriers&token={y}` on your shop you are being returned the Carriers Back Office page.
```
Comment (kpodemski) [inline src/content/1.7/development/components/router-dispatcher.md:40]:
```suggestion
When you visit `/admin{x}/index.php?controller=AdminCarriers&token={y}` on your shop you are being returned to the Carriers Back Office page.
```

File: src/content/1.7/development/components/router-dispatcher.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+What happened behind the scenes is that
```
Comment (kpodemski) [inline src/content/1.7/development/components/router-dispatcher.md:42]:
```suggestion
What happened behind the scenes is that:
```

File: src/content/1.7/development/components/router-dispatcher.md [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+1. Your HTTP request hit the `/admin-{xxx}/index.php` Back Office routing file. This file detected no Symfony route matches your request so it relied on the _Dispatcher_ to handle your query.
+2. The Dispatcher found the right Back Office controller ; the `AdminCarriersController`
+3. The Admin controllers use a generic system to choose what Smarty template use. Generic templates to display forms and listings are available and the controller provides the structure configuration (for example it controls what columns and rows are displayed).
+
+### Migrated Back Office routing
+
+When you visit `/admin-{xxx}/index.php/configure/shop/preferences/preferences?_token={yyy}` on your shop you are being returned the Preferences Back Office page.
```
Comment (kpodemski) [inline src/content/1.7/development/components/router-dispatcher.md:50]:
```suggestion
When you visit `/admin-{xxx}/index.php/configure/shop/preferences/preferences?_token={yyy}` on your shop you are being returned to the Preferences Back Office page.
```

File: src/content/1.7/development/components/router-dispatcher.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+What happened behind the scenes is that
```
Comment (kpodemski) [inline src/content/1.7/development/components/router-dispatcher.md:52]:
```suggestion
What happened behind the scenes is that:
```

File: src/content/1.7/development/components/router-dispatcher.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+1. Your HTTP request hit the `/admin-{xxx}/index.php` Back Office routing file. This file booted the Symfony the kernel to handle your request.
```
Comment (kpodemski) [inline src/content/1.7/development/components/router-dispatcher.md:54]:
```suggestion
1. Your HTTP request hit the `/admin-{xxx}/index.php` Back Office routing file. This file booted the Symfony kernel to handle your request.
```

File: src/content/1.7/development/components/router-dispatcher.md [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+2. Symfony found the right Back controller, matching your URL ; the `PreferencesController`
```
Comment (kpodemski) [inline src/content/1.7/development/components/router-dispatcher.md:55]:
```suggestion
2. Symfony found the right Back Office controller, matching your URL, the `PreferencesController`
```

File: src/content/1.7/development/components/router-dispatcher.md [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+3. The PreferencesController calls the `@PrestaShop/Admin/Configure/ShopParameters/preferences.html.twig` Twig view to be rendered, populated with PHP objects
+
+If you send another HTTP request, the same sequence will happen
+- Symfony finds the right Back controller
```
Comment (kpodemski) [inline src/content/1.7/development/components/router-dispatcher.md:59]:
```suggestion
- Symfony finds the right Back Office controller
```

File: src/content/1.7/development/components/router-dispatcher.md (hunk 5 above)
Comment (matks) [reply src/content/1.7/development/components/router-dispatcher.md:40]: We return a HTML page, not `to a HTML page` ? 🤔 

I think my statement is confusing.
Maybe
```suggestion
When you visit `/admin{x}/index.php?controller=AdminCarriers&token={y}` on your shop, PrestaShop returns the Carriers Back Office HTML page.
```

File: src/content/1.7/development/components/router-dispatcher.md (hunk 7 above)
Comment (matks) [reply src/content/1.7/development/components/router-dispatcher.md:50]: Same ad above

File: src/content/1.7/development/components/router-dispatcher.md (hunk 5 above)
Comment (kpodemski) [reply src/content/1.7/development/components/router-dispatcher.md:40]: yep, sounds better

File: src/content/1.7/development/components/router-dispatcher.md [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+# Router and Dispatcher
+
+The legacy Dispatcher and the Symfony Router are in charge of parsing incoming HTTP requests and finding the right Controller to return HTTP response.
+
+### Front office routing
+
+The Front office controllers lie in the `controllers/front` directory.
+
+The Dispatcher match a given HTTP request with a Front Office controller.
```
Comment (Progi1984) [inline src/content/1.7/development/components/router-dispatcher.md:14]:
```suggestion
The Dispatcher matches a given HTTP request with a Front Office controller.
```

File: src/content/1.7/development/components/router-dispatcher.md (hunk 7 above)
Comment (matks) [reply src/content/1.7/development/components/router-dispatcher.md:50]:
```suggestion
When you visit `/admin-{xxx}/index.php/configure/shop/preferences/preferences?_token={yyy}` on your shop PrestaShop returns the Preferences Back Office HTML page.
```

File: src/content/1.7/development/components/router-dispatcher.md [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+When you visit `/` on your shop you are being returned the Homepage, unless you have your shop installed in a subdirectory, then `/your_subdirectory` is your Homepage.
```
Comment (Progi1984) [inline src/content/1.7/development/components/router-dispatcher.md:16]:
```suggestion
When you visit `/` on your shop, you are being returned the Homepage unless you have your shop installed in a subdirectory, then `/your_subdirectory` is your Homepage.
```

File: src/content/1.7/development/components/router-dispatcher.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+What happened behind the scenes is that:
+
+1. Your HTTP request hit the `index.php` main routing file that relies on the _Dispatcher_
```
Comment (Progi1984) [inline src/content/1.7/development/components/router-dispatcher.md:20]:
```suggestion
1. Your HTTP request hits the `index.php` main routing file that relies on the _Dispatcher_
```

File: src/content/1.7/development/components/router-dispatcher.md [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+## Back office routing
+
+The Back Office is split between the legacy part and the migrated part.
+
+The legacy backend relies on controllers you can find in the `controllers/admin` directory.
+
+The Symfony backend relies on controllers you can find in `src/PrestaShopBundle/Controller/Admin`.
+
+The Back Offie routing match a given HTTP request with a Controller.
```
Comment (Progi1984) [inline src/content/1.7/development/components/router-dispatcher.md:36]:
```suggestion
The Back Offie routing matches a given HTTP request with a Controller.
```

File: src/content/1.7/development/components/router-dispatcher.md [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+
+### Legacy Back Office routing
+
+When you visit `/admin{x}/index.php?controller=AdminCarriers&token={y}` on your shop, PrestaShop returns the Carriers Back Office HTML page.
+
+What happened behind the scenes is that:
+
+1. Your HTTP request hit the `/admin-{xxx}/index.php` Back Office routing file. This file detected no Symfony route matches your request so it relied on the _Dispatcher_ to handle your query.
```
Comment (Progi1984) [inline src/content/1.7/development/components/router-dispatcher.md:44]:
```suggestion
1. Your HTTP request hits the `/admin-{xxx}/index.php` Back Office routing file. This file detected no Symfony route matches your request so it relied on the _Dispatcher_ to handle your query.
```

File: src/content/1.7/development/components/router-dispatcher.md [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+2. The Dispatcher found the right Back Office controller ; the `AdminCarriersController`
```
Comment (Progi1984) [inline src/content/1.7/development/components/router-dispatcher.md:45]:
```suggestion
2. The Dispatcher found the right Back Office controller - the `AdminCarriersController`
```

File: src/content/1.7/development/components/router-dispatcher.md [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+3. The Admin controllers use a generic system to choose what Smarty template use. Generic templates to display forms and listings are available and the controller provides the structure configuration (for example it controls what columns and rows are displayed).
+
+### Migrated Back Office routing
+
+When you visit `/admin-{xxx}/index.php/configure/shop/preferences/preferences?_token={yyy}` on your shop PrestaShop returns the Preferences Back Office HTML page.
+
+What happened behind the scenes is that:
+
+1. Your HTTP request hit the `/admin-{xxx}/index.php` Back Office routing file. This file booted the Symfony kernel to handle your request.
```
Comment (Progi1984) [inline src/content/1.7/development/components/router-dispatcher.md:54]:
```suggestion
1. Your HTTP request hits the `/admin-{xxx}/index.php` Back Office routing file. This file booted the Symfony kernel to handle your request.
```

**Review by Progi1984 — COMMENTED** [review]
Some feedbacks

File: src/content/1.7/development/components/router-dispatcher.md (hunk 14 above)
Comment (matks) [reply src/content/1.7/development/components/router-dispatcher.md:20]: No, this is past tense of `hit` verb, not present. And past tense does not require `s`
The full 1 - 2 - 3 list is written using past tense

File: src/content/1.7/development/components/router-dispatcher.md (hunk 15 above)
Comment (matks) [reply src/content/1.7/development/components/router-dispatcher.md:36]: I saw I also miss the `c` in `Offie` 😆 
```suggestion
The Back Office routing matches a given HTTP request with a Controller.
```

File: src/content/1.7/development/components/router-dispatcher.md (hunk 16 above)
Comment (matks) [reply src/content/1.7/development/components/router-dispatcher.md:44]: Past tense as above

File: src/content/1.7/development/components/router-dispatcher.md (hunk 18 above)
Comment (matks) [reply src/content/1.7/development/components/router-dispatcher.md:54]: Pas tense as above

Comment (PierreRambaud) [reply src/content/1.7/development/components/router-dispatcher.md:54]: It's past tense `This file booted `

**Review by Progi1984 — APPROVED** [review]

**Pull request merged**